### PR TITLE
add toastr, loader and using swr for custom hooks and caching

### PR DESCRIPTION
**Pull request by Walaa-Volidis**
add toastr, loader and using swr for custom hooks and caching

### CODE REVIEW

File: src/app/hooks/useTasks.ts [hunk 1]
```diff
@@ -1,57 +1,84 @@
...
-    if (userId) {
-      fetchTasks();
-    }
-  }, [userId]);
+const fetcher = async (url: string) => {
+  const response = await fetch(url);
+  if (!response.ok) {
+    throw new Error("Failed to fetch tasks");
+  }
+  return response.json();
```
Comment (nmaswood) [inline src/app/hooks/useTasks.ts:10]: **Suggestion: zod parse the response here and return the exact you are expecting**

File: src/app/hooks/useTasks.ts [hunk 2]
```diff
@@ -1,57 +1,84 @@
...
+    fetcher
+  );
+  const [isAdding, setIsAdding] = useState(false);
 
   const addTask = async (formData: FormData) => {
+    setIsAdding(true);
     try {
-      const response = await fetch("/api/createTask", {
+      const formDataTask = {
+        title: formData.get("title") as string,
```
Comment (nmaswood) [inline src/app/hooks/useTasks.ts:24]: **Suggestion: Remove the as string cast and just use zod instead**

File: src/app/hooks/useTasks.ts [hunk 3]
```diff
@@ -1,57 +1,84 @@
...
+
       const newTask = await response.json();
-      setTasks((prevTasks) => [...prevTasks, newTask]);
+
+      mutate("/api/list-tasks");
     } catch (error) {
       console.error("Failed to add task:", error);
+      mutate("/api/list-tasks");
+    } finally {
+      setIsAdding(false);
```
Comment (nmaswood) [inline src/app/hooks/useTasks.ts:49]: **Question: do you need this? can't you just use useMutate useSwr's loading property?**

File: src/app/hooks/useTasks.ts [hunk 4]
```diff
@@ -1,57 +1,84 @@
...
     }
   };
 
   const deleteTask = async (id: string) => {
     try {
-      const response = await fetch(`/api/deleteTask/${id}`, {
+      mutate(
+        "/api/list-tasks",
+        tasks.filter((task) => task.id !== id),
```
Comment (nmaswood) [inline src/app/hooks/useTasks.ts:57]: **Thought: when you mutate data you can call useSWR#mutate that will blow the cache for other requests that depend on it**

**Review by nmaswood — APPROVED** [review]

File: src/app/hooks/useTasks.ts (hunk 1 above)
Comment (Walaa-Volidis) [reply src/app/hooks/useTasks.ts:10]: done

File: src/app/hooks/useTasks.ts (hunk 2 above)
Comment (Walaa-Volidis) [reply src/app/hooks/useTasks.ts:24]: done

File: src/app/hooks/useTasks.ts (hunk 3 above)
Comment (Walaa-Volidis) [reply src/app/hooks/useTasks.ts:49]: yes its not needed, mutate its already loading

**Pull request merged**